custom/icds_reports/data_pull/data_pulls.py

```diff
@@ -193,3 +203,86 @@ def _dump_consolidated_data(result):
+
+
+class VHSNDMonthlyReport(MonthBasedDataPull):
+    slug = "vhsnd_monthly_report"
+    name = "VHSND monthly report"
+    queries = [
+        VHSNDMonthlyCount
+    ]
+
+    def post_run(self, data_files):
+        state_results = self._consolidate_data(data_files)
+        return self._dump_consolidated_data(state_results)
+
+    def _consolidate_data(self, data_files):
+        result = defaultdict(dict)
+        for filename, filestream in data_files.items():
+            filestream.seek(0)
+            reader = csv.DictReader(filestream)
+            for row in reader:
+                state_name = row['state_name']
+                vhsnd_date = parse_date(row['vhsnd_date_past_month']).strftime('%d/%m/%Y')
+                data_key = (row['state_name'], row['district_name'], row['block_name'],
+                            row['supervisor_name'], row['awc_name'])
+                if data_key in result[state_name]:
+                    result[state_name][data_key].append(vhsnd_date)
+                else:
+                    result[state_name][data_key] = [vhsnd_date]
+
+        return self._format_consolidated_data(result)
+
+    def _format_consolidated_data(self, result):
+        # constructing headers
+        headers, dates = self._get_headers_and_dates()
+        # setting up workbooks for states
+        state_results = self._setup_state_results_filestream(headers, result)
+        # populating excel data
+        for state_name, all_details in result.items():
+            for details, vhsnd_dates in all_details.items():
+                awc_row = list(copy(details))
+                total_count = 0
+                for a_date in dates:
+                    if a_date in vhsnd_dates:
+                        awc_row.append(1)
+                        total_count += 1
+                    else:
+                        awc_row.append('')
+                awc_row.append(total_count)
+                state_results[state_name].append(awc_row)
+        return {state_name: state_ws.parent for state_name, state_ws in state_results.items()}
+
+    def _get_headers_and_dates(self):
+        self.month_date = datetime.datetime.strptime(self.month, '%Y-%m-%d')
+        num_days = calendar.monthrange(self.month_date.year, self.month_date.month)[1]
+        days = [datetime.date(self.month_date.year, self.month_date.month, day) for day in range(1, num_days + 1)]
+
+        dates = [day.strftime('%d/%m/%Y') for day in days]
+
+        headers = ['State', 'District', 'Block', 'Sector', 'AWC']
+        headers.extend(dates)
+        headers.append('Grand Total')
+        return headers, dates
+
+    def _setup_state_results_filestream(self, headers, result):
+        state_results = defaultdict(dict)
+        for state_name in result.keys():
+            # constructing and mapping writers to state names
+            if state_name in state_results:
+                raise DuplicateStateResult("%s is twice in results" % state_name)
```

Comment:
result is a dict? how would it have same state name twice? i.e multiple keys with same name twice?

Reply:
That was my suggestion and what you are saying sounds correct to me. If that check is needed it should be moved to a better place.

Reply:
Yes. the check seems unnecessary. It will from the earlier approach where the workbook construction happening on the data population loop. removed it